src/fairseq2/nn/transformer/shaw_attention.py

```diff
@@ -0,0 +1,171 @@
+
+from fairseq2.nn.embedding import Embedding, StandardEmbedding
+from fairseq2.nn.transformer.attention import SDPA
+from fairseq2.typing import DataType, Device, finaloverride
+
+
+@final
+class ShawRelativePositionSDPA(SDPA):
+    """Computes relative position scaled dot-product attention
+    as described in :cite:t:`https://doi.org/10.48550/arxiv.1803.02155`."""
+
+    model_dim: int
+    num_heads: int
+    max_left_rel_pos: int
+    max_right_rel_pos: Optional[int]
+    rel_k_embed: Embedding
+    rel_v_embed: Optional[Embedding]
+
+    def __init__(
+        self,
+        model_dim: int,
+        num_heads: int,
+        max_left_rel_pos: int,
+        *,
+        max_right_rel_pos: Optional[int] = None,
+        use_rel_pos_values: bool = False,
+        attn_dropout_p: float = 0.0,
+        device: Optional[Device] = None,
+        dtype: Optional[DataType] = None,
+    ) -> None:
+        """
+        :param model_dim:
+            The dimensionality of the model.
+        :param: num_heads:
+            The number of attention heads.
+        :param: max_left_rel_pos:
+            The left clipping value for relative positions.
+        :param: max_right_rel_pos:
+            The right clipping value for relative positions.
+        :param: use_rel_pos_values:
+            If True, also uses relative position values to compute relative attention.
+        :param attn_dropout_p:
+            The dropout probability on attention weights.
+        """
+        super().__init__(attn_dropout_p=attn_dropout_p)
+
+        if model_dim % num_heads != 0:
+            raise ValueError(
+                f"`model_dim` must be a multiple of `num_heads` ({num_heads}), but is {model_dim} instead."
+            )
+
+        self.model_dim = model_dim
+        self.num_heads = num_heads
+
+        head_dim = model_dim // num_heads
+
+        self.max_left_rel_pos = max_left_rel_pos
+        self.max_right_rel_pos = (
+            max_right_rel_pos if max_right_rel_pos is not None else max_left_rel_pos
+        )
+        num_pos = self.max_left_rel_pos + 1 + self.max_right_rel_pos
+
+        self.rel_k_embed = StandardEmbedding(
+            num_pos, head_dim, device=device, dtype=dtype
+        )
+
+        if use_rel_pos_values:
+            self.rel_v_embed = StandardEmbedding(
+                num_pos, head_dim, device=device, dtype=dtype
+            )
+        else:
+            self.register_module("rel_v_embed", None)
+
+        self.reset_parameters()
+
+    def reset_parameters(self) -> None:
+        """Reset the parameters and buffers of the module."""
+
+        assert isinstance(self.rel_k_embed.weight, Tensor)
+        nn.init.xavier_uniform_(self.rel_k_embed.weight)
+
+        if self.rel_v_embed is not None:
+            assert isinstance(self.rel_v_embed.weight, Tensor)
+            nn.init.xavier_uniform_(self.rel_v_embed.weight)
+
+    def rel_pos_indices(self, seq_len: int, device: Device) -> Tensor:
```

Comment:
Nit: 1) ideally this should be a private function `def _rel_pos_indices`, 2) for readability I would prefer to have it lexically defined below `forward()`.